### DISCUSSION TITLE
AttributeError for OPENAI_API_KEY import

### DISCUSSION THREAD

**Discussion by hmesfin** [cookiecutter/cookiecutter-django#4845] [Q&A]
Hi All,

Searched discussions for similar issue, but could not find anything related. 

I'm having problem with importing and using an OPENAI_API_KEY  stored in my  **.envs/.production/django** file. 

I have the following in the **production.py**

```OPENAI_API_KEY = env('OPENAI_API_KEY')```

and this in the .production file:

```OPENAI_API_KEY=sk-<some_api_key>```

and calling it like such in my **utils.py**:
```
api_key = settings.OPENAI_API_KEY

client = OpenAI(api_key=api_key)
```

I've also tried ```client = OpenAI(api_key)``` and got the same result.

The result: ```AttributeError: 'Settings' object has no attribute 'OPENAI_API_KEY'``` when I try to build on the production server. The same setup worked just fine locally. 

I am running both local and production via Docker.

Stuck on this for the last few hours, and I'd appreciate any pointers.

**Comment by foarsitter** [comment]
Your key is missing in production. You can either add a default to the variable in production.py or export the correct variable. 

With a default="" your production will come to life but your openai wont work.

**Comment by hmesfin** [comment]
I did that, still no life. Still the same error. It doesn't appear as though the settings entry from production.py is being loaded at all and the default nor the OPENAI_API_KEY read. Puzzling. I could only imagine a typo doing this, but I have gone through a fine toothcomb making sure there are none. The exact same settings entry is read from local.py, and openai works just fine in local dev.

**Reply by foarsitter** [reply]
Did you rebuild your image? Did you restart your production container? Did you made any modifications for production?

**Reply by browniebroke** [reply]
The settings module is selected using the `DJANGO_SETTINGS_MODULE` env variable: 

https://github.com/cookiecutter/cookiecutter-django/blob/d4ff2e4adfe0d2a481ab3d0049bab393a913b659/%7B%7Bcookiecutter.project_slug%7D%7D/.envs/.production/.django#L4C1-L4C50

**Comment by browniebroke** [comment]
Where are you getting this error: locally or on production? You're saying that you defined the setting in `production.py`, but you should probably define it in `base.py`. Here is why: we don't use `production.py` unless it's on prod. For example, If your setting is defined only in `production.py` and you use your `utils.py` locally, Django will look at `local.py` (which itself look at `base.py`) for a variable called `OPENAI_API_KEY`, but there is none.

UPDATE: saw your last answer after posting this

**Reply by hmesfin** [reply]
Not having much luck with adding the api key env variable to the `base.py` file, either.

Below is the entire error output:

Hope that gives you more info to go with, and I appreciate all the responses!

```=> ERROR [python-run-stage 24/24] RUN DATABASE_URL=""   CELERY_BROKER_URL=""   DJANGO_SETTINGS_MODULE="config  1.3s
------
 > [python-run-stage 24/24] RUN DATABASE_URL=""   CELERY_BROKER_URL=""   DJANGO_SETTINGS_MODULE="config.settings.test"   python manage.py compilemessages:
1.210 /usr/local/lib/python3.12/site-packages/environ/environ.py:608: UserWarning: Engine not recognized from url: {'NAME': '', 'USER': '', 'PASSWORD': '', 'HOST': '', 'PORT': '', 'ENGINE': ''}
1.210   warnings.warn(f'Engine not recognized from url: {config}')
1.219 Traceback (most recent call last):
1.219   File "/usr/local/lib/python3.12/site-packages/environ/environ.py", line 388, in get_value
1.219     value = self.ENVIRON[var_name]
1.219             ~~~~~~~~~~~~^^^^^^^^^^
1.220   File "<frozen os>", line 685, in __getitem__
1.220 KeyError: 'OPENAI_API_KEY'
1.220
1.220 The above exception was the direct cause of the following exception:
1.220
1.220 Traceback (most recent call last):
1.220   File "/app/manage.py", line 31, in <module>
1.220     execute_from_command_line(sys.argv)
1.220   File "/usr/local/lib/python3.12/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
1.220     utility.execute()
1.221   File "/usr/local/lib/python3.12/site-packages/django/core/management/__init__.py", line 436, in execute
1.221     self.fetch_command(subcommand).run_from_argv(self.argv)
1.221   File "/usr/local/lib/python3.12/site-packages/django/core/management/base.py", line 412, in run_from_argv
1.222     self.execute(*args, **cmd_options)
1.222   File "/usr/local/lib/python3.12/site-packages/django/core/management/base.py", line 458, in execute
1.222     output = self.handle(*args, **options)
1.222              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1.223   File "/usr/local/lib/python3.12/site-packages/django/core/management/commands/compilemessages.py", line 90, in handle
1.224     basedirs.extend(settings.LOCALE_PATHS)
1.224                     ^^^^^^^^^^^^^^^^^^^^^
1.225   File "/usr/local/lib/python3.12/site-packages/django/conf/__init__.py", line 102, in __getattr__
1.226     self._setup(name)
1.226   File "/usr/local/lib/python3.12/site-packages/django/conf/__init__.py", line 89, in _setup
1.226     self._wrapped = Settings(settings_module)
1.226                     ^^^^^^^^^^^^^^^^^^^^^^^^^
1.228   File "/usr/local/lib/python3.12/site-packages/django/conf/__init__.py", line 217, in __init__
1.228     mod = importlib.import_module(self.SETTINGS_MODULE)
1.229           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1.230   File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
1.230     return _bootstrap._gcd_import(name[level:], package, level)
1.231            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1.232   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
1.232   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
1.232   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
1.232   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
1.232   File "<frozen importlib._bootstrap_external>", line 995, in exec_module
1.232   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
1.232   File "/app/config/settings/test.py", line 5, in <module>
1.232     from .base import *  # noqa
1.232     ^^^^^^^^^^^^^^^^^^^
1.232   File "/app/config/settings/base.py", line 364, in <module>
1.232     OPENAI_API_KEY = env("OPENAI_API_KEY")
1.233                      ^^^^^^^^^^^^^^^^^^^^^
1.233   File "/usr/local/lib/python3.12/site-packages/environ/environ.py", line 199, in __call__
1.233     return self.get_value(
1.233            ^^^^^^^^^^^^^^^
1.233   File "/usr/local/lib/python3.12/site-packages/environ/environ.py", line 392, in get_value
1.233     raise ImproperlyConfigured(error_msg) from exc
1.240 django.core.exceptions.ImproperlyConfigured: Set the OPENAI_API_KEY environment variable
------
Dockerfile:91
--------------------
  90 |
  91 | >>> RUN DATABASE_URL="" \
  92 | >>>   CELERY_BROKER_URL="" \
  93 | >>>   DJANGO_SETTINGS_MODULE="config.settings.test" \
  94 | >>>   python manage.py compilemessages
  95 |
--------------------
ERROR: failed to solve: process "/bin/sh -c DATABASE_URL=\"\"   CELERY_BROKER_URL=\"\"   DJANGO_SETTINGS_MODULE=\"config.settings.test\"   python manage.py compilemessages" did not complete successfully: exit code: 1
ERROR: Service 'django' failed to build : Build failed```

**Reply by browniebroke** [reply]
You need to either set the default value in settings Pytjon file (`OPENAI_API_KEY = env('OPENAI_API_KEY', default=None)`) or set it when we run compilemessage in the Dockerfile (like `DATABASE_URL`):

https://github.com/cookiecutter/cookiecutter-django/blob/474b82b5d3fa67accc3f927e7b1549eadbd2773f/%7B%7Bcookiecutter.project_slug%7D%7D/compose/production/django/Dockerfile#L124-L129
